### DISCUSSION TITLE
Divergent Part Calculation

### DISCUSSION THREAD

**Discussion by wedelfach** [FeynCalc/feyncalc#91] [Q&A] [answered]
Hallo,

I'm calculating a divergent part of the amputated amplitudes generated by FA. The amplitudes originate from one-loop diagrams in some efficient gauge theory. Currently my program converts amplitudes from FC to FA,  contracts and simplifies them. Then in applies: TID, ToPaVe, and PaVeUVPart. All this works correctly, but TID and PaVeUVPart are not very fast. I have to admit, that may expressions is rather long (~1000 terms) but the integrals are not very sophisticated. I'd like to ask, is there any way to speed up this procedure? Or maybe there is an alternative way to exctarct UV part using FeynCalc? I'm interested in purely symbolic answer.

I'm using Mathematica 11.0 (Student Edition), FeynCalc 9.3.1 and FeynArts 3.11.

**Comment by vsht** [comment, marked as answer]
Is TID being used with the UsePaveBasis option? PaVeUVPart works on arbitrary PaVe-functions,
not just the scalar ones.

**Reply by wedelfach** [reply]
No, I haven't used  the "UsePaveBasis" option yet. Is it more efficient than using TID and then ToPaVE? And how to express amplitude in terms of non-scalar Passarano-Veltman functions? As far as I know ToPaVe works only for  scalar integrals.

**Reply by vsht** [reply]
Regarding the efficiency, cf. Sec. 4.1 in 2001.04407.

Consider e.g.
```
int = FVD[p, mu] FVD[p, nu] FVD[p, rho] FVD[p, 
   si] FAD[{p, m0}, {p + q1, m1}, {p + q2, m2}, {p + q3, m3}]
```
You don't really want to reduce it to A0, B0, C0 and D0, since this creates huge
formulas. If you only care about the UV-pole,

```
res = TID[int, p, UsePaVeBasis -> True, ToPaVe -> True]
res // PaVeUVPart
```
returns the result within seconds.

**Reply by wedelfach** [reply]
Thanks a lot, I have red section 4.1 carefully and made some numerical comparison myself. Of course, the method you suggested is much faster. But I have one more question. Does TID group terms proportional to the same loop integrals or should I do it before using TID? Is there a simple way for doing that, or one should simply use the collect command in appropriate way?

**Reply by vsht** [reply]
TID does that automatically for performance reasons, so you don't have to do it yourself.

**Reply by wedelfach** [reply]
I have introduced suggested modification and it has significantly reduced time of divergent part computation. But unfortunately divergent part calculation still takes too long. I believe that this is because I'm dealing with a problem with generic dynamics (all external momenta are non-zero and different). Is there a way to further speed up the whole procedure? Should I expand in external momenta, so that all divergent integrals will not depend on them? Or maybe I should recompile or parallelise the procedure?

**Reply by vsht** [reply]
Well, in general, Mathematica has intrinsic limitations on the number of terms it can handle efficiently. In my experience, once you have more than 30K-40K terms everything becomes too slow, no matter how efficiently you try to organize the computation.

Almost all of the symbolic functions are not compilable, so this won't help here.

Simple parallelization in terms of splitting the diagrams into groups and processing them on separate Mma kernels may 
help, provided that you run your jobs on a computer cluster with many cores. If you are trying to handle everything
on a dual/quad core laptop, then the gain from parallelizing will not be so great.

At the end of the day, you can still try to move the most demanding parts of the calculation to FORM. That, of course, requires
much more work than doing something in FeynCalc, but you also gain quit some performance. But even FORM can be too
slow sometimes and this is why clusters are the way to go in such cases: Even if a single diagram requires 6h of CPU time,
I can still run X diagrams in parallel, meaning that my 4*X diagrams will finish within one day.

PS "Too long" is a very relative statement. For me anything longer than 10 minutes in FeynCalc is already too long so that
I would rather switch to FORM for the given problem. But I also know people who had some FeynCalc amplitude calculations
running for a week and said that it was "fine for them" afterwards.

**Reply by wedelfach** [reply]
Thank You once again, now I have better view. By the way is it possible to group terms proportional to the same loop integrals (exactly as TID is doing) but without performing any other operations?

**Reply by vsht** [reply]
Sure. Consider e.g.

```
lis = Table[Unique[], {i, 1, 6}];
fads = {FAD[{q, m0}], FAD[{q + p1, m1}], FAD[{q + p2, m2}], 
   FAD[{q + p3, m3}], FAD[{q + p4, m4}]};
makeNum[] := 
  FCI[Times @@ Thread[FVD[q, lis[[1 ;; RandomInteger[6]]]]]];
makeDen[] := 
  FeynAmpDenominatorCombine[
   Times @@ RandomSample[fads, RandomInteger[{1, 5}]]];
makePref[] := 
 Sum[RandomInteger[10] x^RandomInteger[10], {i, 1, 5}]/
  Sum[RandomInteger[10] x^RandomInteger[10], {i, 1, 5}]
amp = Sum[makePref[] makeNum[] makeDen[], {i, 1, 100}];
```

Then one could start collecting things in multiple ways

```
AbsoluteTiming[
 ampCollected = 
   Collect2[amp, FeynAmpDenominator, q, Factoring -> {Factor, 5000}, 
    TimeConstrained -> 3];]
```

```
AbsoluteTiming[
 ampCollected = 
   Collect2[amp, FeynAmpDenominator, Factoring -> {Factor, 5000}, 
    TimeConstrained -> 3];]
```

The Factoring and TimeConstrained option tell Collect2 not to bother with factoring the prefactors if the LeafCount is higher than 5000 or if the operation requires more than 3 seconds. These are my empiric defaults that seem to fit most QFT calculations. The reason for doing this is that sometimes Mma's Factor gets too confused with complicated prefactors and wastes too much time on them. In Mma 12.2 WRI  fixed some Factor bugs that I discovered there, so Factor is sort of "cool again". On older Mma versions I'd probably use FC's Factor2 instead.

If you need to temporarily hide the complexity by looking only at certain parts of the expression, use IsolateNames and IsolateFast (and FRH when you want to unisolate)

```
ampIsolated = 
 Collect2[amp, FeynAmpDenominator, q, IsolateNames -> KK, 
  Factoring -> False, IsolateFast -> True]


ampIsolated[[1]]
% // FRH
```

There are also more specific functions for extracting unique loop integrals from the given amplitude that were introduced in
Sec. 3.4 of 1601.01167. So 

FCLoopExtract[amp, {q}, loopInt]

is also a possible choice. If you need to work with amplitudes that contain not only FAD-integrals but also PaVe-Integrals, use
FeynCalc`Package`PaVeHeadsList for the list of all possible PaVe-heads.

**Reply by wedelfach** [reply]
Thank You very much. I found out that the code works also almost 4 times faster if it is called directly from console.

**Reply by vsht** [reply]
Can it be that you are entering your commands without a ";" in the front-end?

The point is that rendering large expressions in the front-end can be very time consuming.
It is also easy to construct cases where the actual evaluation proceeds much faster than the rendering afterwards
E.g.

```
AbsoluteTiming[res = Table[i^2, {i, 1, 10^8}];]
```
vs
```
AbsoluteTiming[res = Table[i^2, {i, 1, 10^8}]]
```
Here you loose a lot of time just because you want to see the intermediate expression
(which is anyhow too large to deduce anything useful out of it).

The front-end is undoubtedly a useful thing, but one has to use it wisely. BTW, it is also
easy to ruin all the performance gains in FORM, by putting a lot of `.sort` commands here
and there thus forcing the compiler to "simplify" the whole expression instead of looking at
it term-by-term as FORM usually does.

Large-scale computations are not just about a powerful language or a lot of CPU power and RAM,
one also has to design the codes in such a way, that computationally expensive operations occur
only when needed, i.e. at right places and nowhere else.

EDIT: Sorry, I actually wanted to have 10^8 instead of 10^6 in my original example. But it seems that
the 10^8 case never terminates on Mma 12.2 (bug?) Anyway, I think that you got my point :)

**Reply by wedelfach** [reply]
Of course, I use semicolons at the end of each line. But thank You for your adivce about the .sort command.

**Comment by vsht** [comment]
Ok then. Although it's still not entirely clear to me, how you get a 4-fold speed increase by using
the kernel instead of the front-end. Provided that the Mma versions are the same in both cases.
